### PR TITLE
Add recover deleted user function

### PR DESCRIPTION
**Pull request by SDSamarasinghe**

### CODE REVIEW

File: src/api/services/TopSpeaker.service.ts [hunk 1]
```diff
@@ -132,6 +132,26 @@ export const deleteTopSpeaker = async (topSpeakerId: string, deletedBy: Schema.T
...
+/**
+recover a past event
+ * @param topSpeakerId @type string
+ */
+export const recoverDeletedTopSpeaker = async (topSpeakerId: string) => {
+	return await TopSpeakerModel.findById(topSpeakerId)
+		.then(async (topSpeakerDetails) => {
+			if (topSpeakerDetails && topSpeakerDetails.deletedAt) {
+				topSpeakerDetails.deletedAt = null as any;
+				topSpeakerDetails.deletedBy = null as any;
```
Comment (rusiruavb) [inline src/api/services/TopSpeaker.service.ts:144]: Is there any specific reason to use `any` in here?

File: src/api/services/TopSpeaker.service.ts [hunk 2]
```diff
@@ -132,6 +132,26 @@ export const deleteTopSpeaker = async (topSpeakerId: string, deletedBy: Schema.T
...
+				return await topSpeakerDetails.save();
+			} else {
+				return null;
```
Comment (rusiruavb) [inline src/api/services/TopSpeaker.service.ts:147]: Do not return `null` to the controller. Add this to `else` part,
```
const errorData = {
   message: "Top speaker information not found",
   dateTime: new Date()
};

throw new Error(JSON.stringify(errorData));
```

File: src/api/controllers/TopSpeaker.controller.ts [hunk 3]
```diff
@@ -113,6 +113,26 @@ export const deleteTopSpeaker = async (request: Request, response: Response, nex
...
+export const recoverDeletedTopSpeaker = async (request: Request, response: Response, next: NextFunction) => {
+	const topSpeakerId = request.params.topSpeakerId;
+
+	if (topSpeakerId) {
+		await TopSpeakerService.recoverDeletedTopSpeaker(request.params.topSpeakerId).then((data) => {
+			request.handleResponse.successRespond(response)(data);
+			next();
+		});
+	} else {
+		request.handleResponse.errorRespond(response)("Top Speaker ID not found");
```
Comment (rusiruavb) [inline src/api/controllers/TopSpeaker.controller.ts:132]: Add `next()` function after handle the response.
```
request.handleResponse.errorRespond(response)("Top Speaker ID not found");
next();
```

**Review by rusiruavb — CHANGES_REQUESTED** [review]

**Comment by rusiruavb** [comment]
There is a conflict in the `package-lock.json` file. Please resolve and update the PR.

File: src/api/services/TopSpeaker.service.ts (hunk 1 above)
Comment (senuravihanjayadeva) [reply src/api/services/TopSpeaker.service.ts:144]: Change it in the interface machn. Reffer Event Interface

**Review by senuravihanjayadeva — CHANGES_REQUESTED** [review]

File: src/api/controllers/TopSpeaker.controller.ts [hunk 4]
```diff
@@ -113,6 +113,27 @@ export const deleteTopSpeaker = async (request: Request, response: Response, nex
...
```
Comment (rusiruavb) [inline src/api/controllers/TopSpeaker.controller.ts:132]: Add date and time to the response.
```
request.handleResponse.errorRespond(response)({ message: "Top Speaker ID not found", dateTime: new Date() });
next();
```

**Review by rusiruavb — CHANGES_REQUESTED** [review]
@SDSamarasinghe I add small comment. Please take a look

File: src/api/routes/index.ts [hunk 5]
```diff
@@ -60,6 +60,7 @@ export default function (app: Express) {
   app.put("/admin/topspeaker/delete/:topSpeakerId", middleware.authenticate, controller.deleteTopSpeaker);
   app.get("/admin/topspeaker/", middleware.authenticate,controller.getAllTopSpeakersForAdmin);
   app.get("/admin/topspeaker/delete/", middleware.authenticate,controller.getDeletedTopSpeakersForAdmin);
+  app.get("/admin/topspeaker/recover/", middleware.authenticate, controller.recoverDeletedTopSpeaker);
```
Comment (senuravihanjayadeva) [inline src/api/routes/index.ts:63]: this should be a "put" request not "get"

**Review by senuravihanjayadeva — CHANGES_REQUESTED** [review]

File: src/api/routes/index.ts (hunk 5 above)
Comment (SDSamarasinghe) [reply src/api/routes/index.ts]: but aiye.. delete function also has a get method neda?

**Review by senuravihanjayadeva — APPROVED** [review]

**Pull request merged**